### ISSUE TITLE
Issue with Lambdas

### ISSUE THREAD

**Issue by GirithGoloth** [lukehutch/fast-classpath-scanner#103]
Hi,

Im running in an issue were if i use lambdas as match processors, he doesnt run and hangs java up.

 i then tried to use inner classes and its working that way. i also tried to do the same scan in a version lower( a lot lower ..) and it worked perfectly. 

Also tried to run a List for each to see if it wasnt my pc freaking out and it works too. the only difference i found in your poms is that the lower is compiled in java 1.8 but the newer is in java 1.8. 

Verbose doesnt say nothing.

Version that hangs: 2.0.10
Version that worked: 1.2

**Comment by lukehutch** [comment]
Thanks for the report. Can you please provide an example?

**Comment by GirithGoloth** [comment]
Hi thanks for the response.

Its attached to the comment.



[classpath_test.zip](https://github.com/lukehutch/fast-classpath-scanner/files/722162/classpath_test.zip)

**Comment by lukehutch** [comment]
Thanks @GirithGoloth, I can replicate the hang using your testcase. I don't understand the hang though. In the Eclipse debugger, if I pause the running thread to examine it, it just says "Waiting for (id = 86)", it doesn't allow me to examine the stacktrace. The fact that the JVM debugger can't even stop the running thread to obtain a full stacktrace indicates something weird is going on.

I compile FCS using Java 1.8, in 1.7 compatibility mode (because I have a lot of users that are not on Java 8 yet). This has always worked before.

The weird thing is that lambdas work in the unit tests in the FastClasspathScanner project itself. So it seems to be a difference with how Maven is compiling FCS for release.

I'll see if I can figure out what Maven is doing differently in the two cases, but it's a very strange bug.

**Comment by lukehutch** [comment]
The symptoms are similar to this OpenJDK bug:

https://bugs.openjdk.java.net/browse/JDK-8145888

**Comment by lukehutch** [comment]
@GirithGoloth I found what's triggering the problem. If you remove the call to `startup()` from the `BeansScanner()` constructor, and then instead call `BeansScanner.getInstance().startUp()`, the code works fine.

This seems to indicate that some part of the Java 8 lambda system is not fully initialized yet if you use lambdas in the static initializer block of the class that contains the `main` method (i.e. before `main` runs).

I assume this is a Java 8 bug? What do you think?

**Comment by GirithGoloth** [comment]
I think that it is related with java 8 definetely since i can actually still use your library fine in other places. Only this class with this combination of steps was with this problem and the weird part is before calling the fastscanner if you create a list and use forEach with a lambda, it works, even if i call parallelStream and also adding, when it hangs no exception is thrown.

Im closing this issue as it doesnt seem related to FastClasspathScanner. Thanks for your replies and keep up the good work.

**Comment by lukehutch** [comment]
@GirithGoloth it hangs somewhere deep in the JVM, so it does seem to be a Java 8 bug. I filed a report at the Oracle website and directed them here, hopefully they will take a look. I guess just make sure that `main` has started running before you call FashClasspathScanner. Thanks for the report!

**Comment by lukehutch** [comment]
@GirithGoloth OK, this is definitely a JRE bug. I created this minimal testcase to illustrate the problem:

```java
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class LambdaBug {

    // Code will hang after "Entering startUp()" if and only if both these are true
    private static final boolean RUN_AS_STATIC_INITIALIZER = true;
    private static final boolean RUN_USING_LAMBDA = true;

    static {
        if (RUN_AS_STATIC_INITIALIZER) {
            startUp();
        }
    }

    private static void startUp() {
        System.out.println("Entering startUp()");
        ExecutorService es = Executors.newSingleThreadExecutor();
        try {
            Callable<Void> callable;
            if (RUN_USING_LAMBDA) {
                callable = () -> {
                    System.out.println("Lambda executed");
                    return null;
                };
            } else {
                callable = new Callable<Void>() {
                    @Override
                    public Void call() throws Exception {
                        System.out.println("Inner class method executed");
                        return null;
                    }
                };
            }
            es.submit(callable).get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
        es.shutdown();
        System.out.println("Exiting startUp()");
    }

    public static void main(String[] args) {
        if (!RUN_AS_STATIC_INITIALIZER) {
            startUp();
        }
        System.out.println("Exiting main");
    }
}
```

**Comment by lukehutch** [comment]
PS to trigger this, you have to be running in another thread, which is why the 1.2 version did not exhibit this problem -- it was single-threaded.

**Comment by lukehutch** [comment]
I got an explanation from David Holmes of Oracle about why this code example causes deadlock -- very interesting:

http://mail.openjdk.java.net/pipermail/core-libs-dev/2017-January/046060.html

I still have questions about whether this can be prevented, or at least detected by FastClasspathScanner (so that FCS doesn't accept Lambda expressions in this context) -- see that conversation thread.

**Comment by lukehutch** [comment]
I discovered this issue doesn't just happen with lambdas, it can happen with anonymous inner classes too, if they access the fields of the class they are defined within. The deadlock is hit as soon as they try to access those fields. (Presumably this also happens with method calls.)

**Comment by GirithGoloth** [comment]
Hi,

Well seems we have to be careful were we use and lambdas and Threads. I hope they somehow create a compilation warning so that we can be careful on doing this kind of stuff, but with what im seeing this is a very hard pattern to track and sometimes it would spam warning in the console which would be undesirable. So for now might well be aware of this situation and how to prevent it .

Thanks for keeping this updated on the problem.

**Comment by lukehutch** [comment]
@GirithGoloth see the discussion on the core-libs-dev mailing list above. The right solution for FastClasspathScanner is to run `MatchProcessors` on the calling thread, so I just refactored the code to do that. Your original code example should work fine now.

The only form of scanning you can't do from a class initializer block now is asynchronous scanning, i.e. calling `scanAsync()` rather than `scan()`, since this will run `MatchProcessors` on a separate thread. FastClasspathScanner will now detect this case (of asynchronous scanning, launched from a class initializer block), and will throw a RuntimeException, refusing to start, in preference to causing a deadlock.

Thanks for this bug report, fixing this was a fascinating journey! This fix will be in version 2.0.13.
